ncurses: add -std=gnu17 to fix gcc15 build error and update to 6.5

graysky2 opened the pull request:
1. Update to 6.5 which required a manual reabase of 01-ncurses-5.6-20080628-kbs.patch
2. Add -std=gnu17 to CFLAGS to avoid a GCC15 error when building with glibc

Without the newly introduced flag, building with libc ends in errors such as:
```
/scratch/union/staging_dir/toolchain-x86_64_gcc-15.1.0_glibc/x86_64-openwrt-linux-gnu/include/c++/15.1.0/cstddef:81:21: error: redefinition of 'struct std::__byte_operand<unsigned char>'
   81 |   template<> struct __byte_operand<unsigned char> { using __type =1. Update to 6.5 which required a manual reabase of 01-ncurses-5.6-20080628-kbs.patch
 byte; };
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/scratch/union/staging_dir/toolchain-x86_64_gcc-15.1.0_glibc/x86_64-openwrt-linux-gnu/include/c++/15.1.0/cstddef:78:21: note: previous definition of 'struct std::__byte_operand<unsigned char>'
   78 |   template<> struct __byte_operand<bool> { using __type = byte; };
      |                     ^~~~~~~~~~~~~~~~~~~~
make[4]: *** [Makefile:438: ../obj_s/cursesp.o] Error 1
```

robimarko commented:
How are you hitting this?

Cause I have GCC15 on the host as well as it being the OpenWrt toolchain and ncurses just builds fine

graysky2 commented:
> How are you hitting this?

I switched to building with glibc from musl and hit it.

neheb commented:
Is this a dependent package? That’s still a C++ header.

graysky2 commented:
Here is my difconfig:
```
CONFIG_TARGET_x86=y
CONFIG_TARGET_x86_64=y
CONFIG_TARGET_x86_64_DEVICE_generic=y
CONFIG_DEVEL=y
CONFIG_TOOLCHAINOPTS=y
CONFIG_BUSYBOX_CUSTOM=y
# CONFIG_BINUTILS_USE_VERSION_2_42 is not set
CONFIG_BINUTILS_USE_VERSION_2_44=y
CONFIG_BINUTILS_VERSION="2.44"
CONFIG_BINUTILS_VERSION_2_44=y
# CONFIG_GCC_USE_VERSION_14 is not set
CONFIG_GCC_USE_VERSION_15=y
CONFIG_GCC_VERSION="15.1.0"
CONFIG_GCC_VERSION_15=y
# CONFIG_KERNEL_WERROR is not set
CONFIG_LIBC="glibc"
CONFIG_LIBC_USE_GLIBC=y
# CONFIG_LIBC_USE_MUSL is not set
CONFIG_STRIP_ARGS="--strip-all"
CONFIG_TARGET_SUFFIX="gnu"
CONFIG_USE_GLIBC=y
CONFIG_USE_STRIP=y
```

graysky2 commented:
I just tried but getting the same error.

graysky2 commented:
I updated the PR, can someone please review my manual reabase of the patch mentioned in the commit log?

neheb commented:
given that the patch is older than 2012, it's questionable if it's even needed. My guess is that it was dropped before ncurses made its release.

neheb commented:
same with some of the others.

Does removing 102-ncurses-5.9-gcc-5.patch fix your issue?

neheb commented:
I think upstream's fix is: https://ncurses.scripts.mit.edu/?p=ncurses.git;a=commit;h=394a1a6cf317912584592e33184ef550e738a4b9

Rather involved...

robimarko commented:
Please split up the update and GNU17 setting into 2 commits.

It does not look easy to backport fix, so until a new ncurses release I am fine with setting GNU17

graysky2 committed:
ncurses: Add -std=gnu17 to CFLAGS to avoid a GCC15 error when building with glibc

Without the newly introduced flag, building with libc ends in errors such the below.
There is an upstream fix[1], but backporting it is not straight forward.

/scratch/union/staging_dir/toolchain-x86_64_gcc-15.1.0_glibc/x86_64-openwrt-linux-gnu/include/c++/15.1.0/cstddef:81:21: error: redefinition of 'struct std::__byte_operand<unsigned char>'
   81 |   template<> struct __byte_operand<unsigned char> { using __type = byte; };
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/scratch/union/staging_dir/toolchain-x86_64_gcc-15.1.0_glibc/x86_64-openwrt-linux-gnu/include/c++/15.1.0/cstddef:78:21: note: previous definition of 'struct std::__byte_operand<unsigned char>'
   78 |   template<> struct __byte_operand<bool> { using __type = byte; };
      |                     ^~~~~~~~~~~~~~~~~~~~
make[4]: *** [Makefile:438: ../obj_s/cursesp.o] Error 1

1. https://ncurses.scripts.mit.edu/?p=ncurses.git;a=commit;h=394a1a6cf317912584592e33184ef550e738a4b9

Signed-off-by: John Audia <therealgraysky@proton.me>
Link: https://github.com/openwrt/openwrt/pull/19335
Signed-off-by: Robert Marko <robimarko@gmail.com>

graysky2 committed:
ncurses: update to 6.5

Update to 6.5.

Removed patch from pre-2012: 101-ncurses-5.6-20080628-kbs.patch

Build system: x86/64
Build-tested: x86/64-glibc
Run-tested: x86/64-glibc

Signed-off-by: John Audia <therealgraysky@proton.me>
Link: https://github.com/openwrt/openwrt/pull/19335
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Ugh, well this broke the host ncurses build:
```
** Building terminfo database, please wait...
Running tic to install /usr/lib64/kitty/terminfo ...

	You may see messages regarding extended capabilities, e.g., AX.
	These are extended terminal capabilities which are compiled
	using
		tic -x
	If you have ncurses 4.2 applications, you should read the INSTALL
	document, and install the terminfo without the -x option.

ncurses 6.5.20240427
"terminfo.tmp", line 24494, terminal 'v3220': /usr/lib64/kitty/terminfo: permission denied (errno 13)
? tic could not build /usr/lib64/kitty/terminfo
make[5]: *** [Makefile:109: install.data] Error 1
```
